wasm/types.go

```diff
@@ -0,0 +1,273 @@
+}
+
+type InvalidTypeConstructorError struct {
+	Wanted int
+	Got    int
+}
+
+func (e InvalidTypeConstructorError) Error() string {
+	return fmt.Sprintf("wasm: invalid type constructor: wanted %d, got %d", e.Wanted, e.Got)
+}
+
+func readFunction(r io.Reader) (FunctionSig, error) {
+	f := FunctionSig{}
+
+	form, err := leb128.ReadVarint32(r)
+	if err != nil {
+		return f, err
+	}
+
+	f.Form = int8(form)
+
+	paramCount, err := leb128.ReadVarUint32(r)
+	if err != nil {
+		return f, err
+	}
+	f.ParamTypes = make([]ValueType, paramCount)
+
+	for i := uint32(0); i < paramCount; i++ {
+		f.ParamTypes[i], err = readValueType(r)
+		if err != nil {
+			return f, err
+		}
+	}
+
+	returnCount, err := leb128.ReadVarUint32(r)
+	if err != nil {
+		return f, err
+	}
+
+	for i := uint32(0); i < returnCount; i++ {
+		vt, err := readValueType(r)
+		if err != nil {
+			return f, err
+		}
+		f.ReturnTypes = append(f.ReturnTypes, vt)
+	}
+
+	return f, nil
+}
+
+// GlobalVar describes the type and mutability of a declared global variable
+type GlobalVar struct {
+	// Type of the value
+	ContentType ValueType
+	Mutable     bool
+}
+
+func readGlobalVar(r io.Reader) (*GlobalVar, error) {
+	g := &GlobalVar{}
+	var err error
+
+	g.ContentType, err = readValueType(r)
+	if err != nil {
+		return nil, err
+	}
+
+	m, err := leb128.ReadVarUint32(r)
+	if err != nil {
+		return nil, err
+	}
+
+	g.Mutable = m == 1
+
+	return g, nil
+}
+
+// Table describes a table in a Wasm module.
+type Table struct {
+	// The type of elements
+	ElementType ElemType
+	Limits      ResizableLimits
+}
+
+func readTable(r io.Reader) (*Table, error) {
+	t := Table{}
+	var err error
+
+	t.ElementType, err = readElemType(r)
+	if err != nil {
+		return nil, err
+	}
+
+	lims, err := readResizableLimits(r)
+	if err != nil {
+		return nil, err
+	}
+
+	t.Limits = *lims
+	return &t, err
+}
+
+type Memory struct {
+	Limits ResizableLimits
+}
+
+func readMemory(r io.Reader) (*Memory, error) {
+	lim, err := readResizableLimits(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Memory{*lim}, nil
+}
+
+// External describes the kind of the entry being imported or exported.
+type External uint8
+
+const (
+	ExternalFunction External = 0
+	ExternalTable    External = 1
+	ExternalMemory   External = 2
+	ExternalGlobal   External = 3
+)
+
+func (e External) String() string {
+	switch e {
+	case ExternalFunction:
+		return "function"
+	case ExternalTable:
+		return "table"
+	case ExternalMemory:
+		return "memory"
+	case ExternalGlobal:
+		return "global"
+	default:
+		return "<unknown external_kind>"
+	}
+}
+func readExternal(r io.Reader) (External, error) {
+	bytes, err := readBytes(r, 1)
+
+	return External(bytes[0]), err
+}
+
+type ResizableLimits struct {
+	// true if the maximum field is present
+	Flags bool
+	// initial length (in units of table elements or wasm pages)
+	Initial uint32
+	// non-nil if specified by flags
+	Maximum *uint32
```

Comment:
> but perhaps Flags will also encode some additional fields of ResizableLimits

Yes, you're right. @dennwc had pointed that out in another review, so this will have to stay.